src/main/scala/exercise2/Stack.scala

```diff
@@ -0,0 +1,41 @@
+package exercise2
+
+import scala.util.Try
+
+trait StackLike[T] {
+  def push(elem: T): StackLike[T]
+
+  def pop(): Try[StackLike[T]]
+
+  def top(): Option[T]
+
+  def isEmpty: Boolean
+
+  def reverse: StackLike[T]
+}
+
+class Stack[T](val elem: Option[T] = None, val tail: Option[Stack[T]] = None)
+    extends StackLike[T] {
+
+  def push(elem: T): StackLike[T] = new Stack(Option(elem), Option(this))
+
+  def pop(): Try[StackLike[T]] = if (isEmpty)
```

Comment:
Very bad! Why bother with `Try` when you throw an exception here anyway?
instead use
```scala
new Failure(new Throwable("Stack is empty"))
```